feat: add liquidity dev docs (#87)
tact-lang/dex: dev-docs/liquidity.md

5. **AMM Pool Mints LP Jettons**  
   The pool mints LP jettons to the depositor and, if necessary, returns any excess assets to the user if the deposit ratio was not exact.

T-Dex follows the Uniswap V2 specification (TODO: add this section and cross-link to it), so liquidity provisioning math is the same.

If it is the first time liquidity is being added to the pool, then `sqrt(leftSideReceived * rightSideReceived)` LP tokens are minted to the depositor.

If it is **not** the first time, minted LP tokens follow this formula:

```tact
liquidityTokensToMint = min(
    muldiv(leftSideReceived, self.totalSupply, self.leftSideReserve - leftSideReceived),
    muldiv(rightSideReceived, self.totalSupply, self.rightSideReserve - rightSideReceived),
);
```

#### Example (Jetton Vault)

```typescript
const payload = createJettonVaultLiquidityDepositPayload(
    liquidityDepositContractAddress,
    /* proofCode, proofData, */ // for advanced use, TODO: add proof link
    minAmountToDeposit,
    lpTimeout,
    payloadOnSuccess,
    payloadOnFailure,
)
const depositLiquidityResult = await jettonWallet.sendTransfer(
    provider,
    sender,

>     toNano("0.05"), // TON for fees

Reviewer: At least 0.1 storage for each invoked contract + gas fees. So something about 0.5 Ton should be ok

    jettonAmount,
    vaultAddress,
    responseAddress,
    null,
    toNano("0.01"),
    payload,
)
```

#### Example (TON Vault)

```typescript
const payload = createTonVaultLiquidityDepositPayload(
    liquidityDepositContractAddress,
    tonAmount,
    payloadOnSuccess,
    payloadOnFailure,
    minAmountToDeposit,
    lpTimeout,
)
// Send TON with this payload to the vault address
const depositLiquidityResult = await wallet.send({
    to: vault.address,

>     value: tonAmount + toNano(0.2), // gas fee

Reviewer: I think here we need 0.5 tons at least

```tact
let successForwardPayload = beginCell()
    .storeBit(false) // Either bit equals 0
    .storeMaybeRef(msg.leftAdditionalParams.payloadOnSuccess)
    .storeMaybeRef(msg.rightAdditionalParams.payloadOnSuccess)
    .endCell()
    .beginParse();
```

- Liquidity deposit is successful, but some funds are refunded due to the constant product invariant; the success payload is attached to the vault payout message (TLB is asset dependent)

If liquidity deposit is successful but one of the assets is refunded, the success payload from that asset's `AdditionalParams` will be attached to the refund payout message **and** both success payloads will be attached to the LP Jetton mint transfer notification.

## Removing Liquidity (Withdrawing)

To withdraw your share, you must burn your LP jettons. The AMM pool will send the corresponding amounts of each asset back to you.

### Step-by-step

1. **Burn LP Jettons**

    - Use your LP jetton wallet to send a burn message with a special payload to the AMM pool.
    - The payload should specify:
        - Minimum amounts of each asset you are willing to receive (to protect against slippage)
        - Timeout
        - Receiver address

>         - (Optional) Callback payload

Reviewer: Mention please that it is only for successful lp withdrawing

`LPWithdrawViaJettonBurn` follows TEP-74 for burning Jettons. However, `customPayload` should serialize to a `LiquidityWithdrawParameters` ref cell. If it is something else or `null`, liquidity withdrawal will be stopped. Also, since the LP Jetton wallet contract can handle bounces on burn from the Jetton minter (which is the AMM pool), all refunds and operation cancels (due to slippage, timeout, or other reasons) are done via throw+bounce.

2. **AMM Pool Processes Withdrawal**
    - The pool calculates the amounts to return based on your share.
    - If the minimums are met, the pool sends payouts from each vault to your address.
    - If not, the transaction is reverted.

If you provide `liquidityWithdrawPayload`, it will be attached to the payout message:

- For TON vault, it will be the whole body (empty if null).
- For Jetton vault, it will be serialized as a reference forward payload in the Jetton transfer notification.

#### Example

```typescript
const withdrawPayload = createWithdrawLiquidityBody(
    minAmountLeft,
    minAmountRight,
    timeout,
    receiver,
    successfulPayload,
)
await lpJettonWallet.sendBurn(
    provider,
    sender,

>     toNano("0.05"), // TON for fees

Reviewer: Not enough tons

# Liquidity management

> This section explains how to add and remove liquidity in T-Dex, enabling users to participate as liquidity providers and earn a share of trading fees.

Author: ```suggestion
This section suggests about ... to add and remove liquidity in T-Dex, enabling users to participate as liquidity providers and earn a share of trading fees.
```

# Liquidity management

This section explains how to add and remove liquidity in T-Dex, enabling users to participate as liquidity providers and earn a share of trading fees.

## Overview

T-Dex uses a two-vault system for each pool (e.g., TON Vault and Jetton Vault). To add liquidity, you must deposit both assets into their respective vaults in the correct ratio. In return, you receive LP (liquidity provider) jettons, which represent your share of the pool. To withdraw liquidity, you burn your LP jettons and receive the underlying assets back.

## Adding Liquidity

### Prerequisites

> - Both vaults (for each asset in the pool) must be deployed and initialized.

Reviewer: And actually we don't need initialized vault. Architecture allows to initialize it on the fly.

But yeah, maybe we don't need to mention it here

- The AMM pool must exist.
- You need the addresses of both vaults and the pool.

Note that throughout T-Dex documentation, the terms **left** and **right** are used to describe asset vaults. For determinism in on-chain operations and predictable addresses, vaults should be sorted and used based on their contract address ordering. (TODO: link vault ordering from vaults docs)

There are two ways one can deploy liquidity deposit contract.

The first one is to deploy liquidity deposit contract with standalone deploy message with state init attached and later only use plain `messages` to communicate liquidity deposit of asset parts.

> The second option is to provide additional fields to the vault's liquidity deposit messages (`part_has_been_deposited#e7a3475f`), so they would deploy liquidity deposit contract themselves.

Reviewer: it's better to mention lp_deposit_part. As part_has_been_deposited is internal interface

> Note, that it is not recommended to use different approach for each one of the vaults - since TON is asynchronous, you can't make sure which liquidity deposit message will come first. So, if you attach state init only to one of the vault messages, it is possible to break an invariant and lose funds.

Reviewer: However, deploying lp provider in separate message and then depositing liquidity is totally fine as transactions in TON are sorted based on their LT.

- `otherAmount` is the amount of the other asset you want to deposit.
- `contractId` is an on-chain salt and can be any number (see the full description later).

2. **Deposit Asset A and Asset B**

    - Send a transfer to each vault (TON or Jetton) with a special payload referencing the Liquidity Deposit contract.
    - The payload must include:
        - The address of the Liquidity Deposit contract
        - The amount to deposit
        - (Optional) Minimum amount to accept, timeout, and callback payloads

For Jetton vaults, use a jetton transfer with a forward payload created by a helper like `createJettonVaultLiquidityDepositPayload`.  
For TON vaults, send a TON transfer with a similar payload.

TLB for adding liquidity:

```tlb
_ minAmountToDeposit:Coins
  lpTimeout:uint32
  payloadOnSuccess:(Maybe ^Cell)
  payloadOnFailure:(Maybe ^Cell) = AdditionalParams;

add_liquidity_part_ton#1b434676
    amountIn:Coins
    liquidityDepositContractData:LiquidityDepositEitherAddress
    additionalParams:AdditionalParams = AddLiquidityPartTon;

add_liquidity_part_jetton#64c08bfc
    liquidityDepositContractData:LiquidityDepositEitherAddress
    additionalParams:AdditionalParams

>     proofType:(##8) {proofType = 0} = AddLiquidityJettonForwardPayload;

Reviewer: let's mention the importance of initialization one more time

    to: vault.address,
    value: tonAmount + toNano(0.5), // gas fee
    bounce: true,
    body: payload,
})
```

### Payload semantics

Similar to [Jetton swap message payloads](./swap.md#payload-semantics), you can attach `payloadOnSuccess` and `payloadOnFailure` to the liquidity deposit message as optional reference cells. These payloads allow you to interact with the protocol on-chain and use them as async callbacks or notifications after the liquidity deposit operation or refund.

If you attach them to the deposit message, one of these payloads (depending on the outcome) will be attached in the vault's `payout` message in case of refund (the TLB of how the asset is delivered after the vault payout is asset-dependent. TODO: add link to the vaults section with payout message structs) or it will be attached to the LP Jetton mint transfer notification (as a reference forward payload, TLB for this follows TEP-74).

**Failure payload** is attached to both payout refund messages when:

- Both deposit values are refunded because the timeout check failed
- Both deposit values are refunded because slippage is too high

**Success payload** is attached to the message when:

> - Liquidity deposit is successful; both successful payloads are attached to the LP Jetton mint transfer notification

Reviewer: - Liquidity deposit is successful; Excesses will be paid with **Success payload**
